server/coordinator/procedure/shard/shard_fsm.go

```diff
@@ -0,0 +1,229 @@
+var (
+	leaderFsmEvent = fsm.Events{
+		{Name: EventPrepareTransferFollower, Src: []string{StateLeader}, Dst: StatePendingFollower},
+		{Name: EventTransferFollower, Src: []string{StatePendingFollower}, Dst: StateFollower},
+		{Name: EventTransferFollowerFailed, Src: []string{StatePendingFollower}, Dst: StateLeader},
+	}
+	leaderFsmCallbacks = fsm.Callbacks{
+		EventPrepareTransferFollower: prepareTransferFollowerCallback,
+		EventTransferFollower:        transferFollowerCallback,
+		EventTransferFollowerFailed:  transferFollowerFailedCallback,
+	}
+)
+
+var (
+	followerFsmEvent = fsm.Events{
+		{Name: EventPrepareTransferLeader, Src: []string{StateFollower}, Dst: StatePendingLeader},
+		{Name: EventTransferLeader, Src: []string{StatePendingLeader}, Dst: StateLeader},
+		{Name: EventTransferLeaderFailed, Src: []string{StatePendingLeader}, Dst: StateFollower},
+	}
+	followerFsmCallbacks = fsm.Callbacks{
+		EventPrepareTransferLeader: prepareTransferLeaderCallback,
+		EventTransferLeader:        transferLeaderCallback,
+		EventTransferLeaderFailed:  transferLeaderFailed,
+	}
+)
+
+type LeaderCallbackRequest struct {
+	Ctx            context.Context
+	C              *cluster.Cluster
+	ActionDispatch dispatch.ActionDispatch
+
+	OldLeaderNode    string
+	OldLeaderShardID uint32
+}
+
+type FollowerCallbackRequest struct {
+	Ctx            context.Context
+	C              *cluster.Cluster
+	ActionDispatch dispatch.ActionDispatch
+
+	NewLeaderNode    string
+	NewLeaderShardID uint32
+}
+
+// NewShardFSM
+/**
+```
+┌────┐                   ┌────┐
+│ RW ├─────────┐         │ R  ├─────────┐
+├────┘         │         ├────┘         │
+│    Leader    ◀─────────│PendingLeader │
+│              │         │              │
+└───────┬──▲───┘         └───────▲─┬────┘
+        │  │                     │ │
+┌────┐  │  │             ┌────┐  │ │
+│ R  ├──▼──┴───┐         │ R  ├──┴─▼────┐
+├────┘         │         ├────┘         │
+│   Pending    ├─────────▶   Follower   │
+│   Follower   │         │              │
+└──────────────┘         └──────────────┘
+```
+*/
+func NewShardFSM(role clusterpb.ShardRole) *fsm.FSM {
+	switch role {
+	case clusterpb.ShardRole_LEADER:
+		leaderShardFsm := fsm.NewFSM(
+			StateLeader,
+			leaderFsmEvent,
+			leaderFsmCallbacks,
+		)
+		return leaderShardFsm
+	case clusterpb.ShardRole_FOLLOWER:
+		followerShardFsm := fsm.NewFSM(
+			StateFollower,
+			followerFsmEvent,
+			followerFsmCallbacks,
+		)
+		return followerShardFsm
+	}
+	return nil
+}
+
+func prepareTransferFollowerCallback(event *fsm.Event) {
+	request := event.Args[0].(*LeaderCallbackRequest)
+	ctx := request.Ctx
+	c := request.C
+	oldLeaderShardID := request.OldLeaderShardID
+
+	// TODO: add ShardRole_PENDING_FOLLOWER enum, replace ShardRole_LEADER to ShardRole_PENDING_FOLLOWER
+	if err := updateShardRole(ctx, c, oldLeaderShardID, clusterpb.ShardRole_FOLLOWER); err != nil {
+		event.Cancel(errors.Wrap(err, EventPrepareTransferFollower))
+	}
```

Comment:
It is enough to keep the TODO, and no need to put wrong codes here.

Reply:
Remove this and just leave TODO will cause CI to fail.